### PR TITLE
fixes #11: allow ES6 unicode identifiers in isIdentifier{Start,Part}

### LINKED ISSUE #11: esutils.code.isIdentifierPart doesn't recognise ES6 IdentifierParts
The only one I'm aware of is `\u{e01d5}`. /cc @mathiasbynens I think we just need to remove these lines from generate-identifier-regex.js:

```js
        .removeRange(0x010000, 0x10FFFF) // remove astral symbols
```

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #11.

### CODE REVIEW

**Comment by Constellation** [comment]
Overall LGTM. Could you add tests?

**Comment by michaelficarra** [comment]
This was a very rough PR. We shouldn't merge because these are not valid in ES5 (I believe). We should probably create two separate interfaces: `isIdentifierStartES6`, etc.

**Comment by mathiasbynens** [comment]
See my comments here: https://github.com/estools/esutils/issues/11#issuecomment-78215168 Let’s follow the spec exactly.

**Comment by michaelficarra** [comment]
I've updated the PR, incorporating @mathiasbynens's changes and adding support for surrogate pairs. This will require a major version bump, as it splits isIdentifier{Start,Part,Name} into ES5 and ES6 variants.

File: tools/generate-identifier-regex.js [hunk 1]
```diff
@@ -34,19 +35,53 @@ var generateES5Regex = function() { // ES 5.1
...
+        .removeRange(0x0, 0x7F); // remove ASCII symbols (esutils-specific)
+    var identifierPart = regenerate(ID_Continue)
+        // Note: `ID_Continue` already includes `Other_ID_Continue`. http://git.io/wRCAfQ
+        .add(Other_ID_Start)
+        .add('\u200C', '\u200D')
+        .removeRange(0x0, 0x7F); // remove ASCII symbols (esutils-specific)
+
+    return {
+        'NonAsciiIdentifierStart': '/' + identifierStart + '/',
+        'NonAsciiIdentifierPart': '/' +identifierPart + '/',
```
Comment (mathiasbynens) [inline tools/generate-identifier-regex.js]: space

Comment (michaelficarra) [inline tools/generate-identifier-regex.js]: Fixed. Thanks.

File: lib/keyword.js [hunk 2]
```diff
@@ -94,33 +94,60 @@
...
+                lowCh = id.charCodeAt(i);
+                if (!(0xDC00 <= lowCh && lowCh <= 0xDFFF)) {
+                    return false;
+                }
+                ch = decodeUtf16(ch, lowCh);
+            }
+            if (!check(ch)) {
                 return false;
             }
+            check = code.isIdentifierPartES6;
```
Comment (mathiasbynens) [inline lib/keyword.js]: Note: `isIdentifierStart` and `isIdentifierPart` still use `String.fromCharCode` which [doesn’t support astral code points correctly](https://mathiasbynens.be/notes/javascript-unicode#code-point-to-symbol). Is there any code that accounts for this? (I don’t see it but I may be missing something.)

Comment (michaelficarra) [inline lib/keyword.js]: You've missed it. They now use `fromCodePoint`.

Comment (mathiasbynens) [inline lib/keyword.js]: Ah, GitHub had hidden the diff for that file, so my `Cmd`+`F`-fu failed. Thanks!

**Comment by michaelficarra** [comment]
Can I get a LGTM so we can release this? It's the last thing blocking the release of the ES6 version of the Shift Parser :grin:.

**Comment by mathiasbynens** [comment]
LGTM

**Comment by michaelficarra** [comment]
@Constellation Can you add me (or at least one other estools member) as an author to each of your npm modules in this organisation? We should increase our [bus factor](http://en.wikipedia.org/wiki/Bus_factor) for these important tools.

**Comment by Constellation** [comment]
@michaelficarra 

OK! I've added you (npm id: michaelficarra) to

- escope
- esutils
- estraverse
- escodegen
- esmangle
- esrecurse
- esshorten

:D

**Pull request merged**